hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java

```diff
@@ -1301,4 +1369,33 @@ public void close() {
     this.heartbeatClient.stop();
     this.txnManager.close();
   }
+
+  private void setWriteTimer(HoodieTable<T, I, K, O> table) {
+    String commitType = table.getMetaClient().getCommitActionType();
+    if (commitType.equals(HoodieTimeline.COMMIT_ACTION)) {
+      writeTimer = metrics.getCommitCtx();
+    } else if (commitType.equals(HoodieTimeline.DELTA_COMMIT_ACTION)) {
+      writeTimer = metrics.getDeltaCommitCtx();
+    }
+  }
+
+  private void tryUpgrade(HoodieTableMetaClient metaClient, Option<String> instantTime) {
+    UpgradeDowngrade upgradeDowngrade =
+        new UpgradeDowngrade(metaClient, config, context, upgradeDowngradeHelper);
+
+    if (upgradeDowngrade.needsUpgradeOrDowngrade(HoodieTableVersion.current())) {
```

Comment:
as per master, this code is available only in SparkEngine. Flink and java does not have this. Even if we wish to unify, I would do it in a separate patch and getting it reviewed by experts who have worked on it. Can you move this to SparkRDDWriteClient for now.

Reply:
I don't think this is accurate -- Flink is also using `UpgradeDowngrade`. Only Java-engine didn't have corresponding helper which i've bootstrapped in here, otherwise code of the `UpgradeDowngrade` itself is engine-agnostic.

Can you please elaborate what your concern around it is? 

Frankly, am not a fan that i had to do budge this refactoring in this PR, but after adding validations into `HoodieMetadataPayload` -- tests started to fail b/c we were not init'ing MT appropriately, so i had to address that holistically across all engines.

Reply:
I am fine. but just that I want to get a stamp it from devs who work in flink and java. Just don't want to miss anything, bcoz, not many times every engine follows same logic. I checked the code and refactoring once again, seems to be fine. 
@danny0405 : we are making some minor refactoring relating to upgradedowngrade invocation for flink engine. just wanted to keep you in the loop.

Reply:
@yihua : we are making some minor refactoring relating to upgradedowngrade invocation for java engine. just wanted to keep you in the loop. Prior to this patch, we did not have any explicit upgrade downgrade. Just wanted to ensure unification does not cause any issues. if you can review it once and stamp it, would be nice.